core/services/run_queue.go

```diff
@@ -34,81 +35,176 @@ type RunQueue interface {
 }
 
 type runQueue struct {
-	workersMutex  sync.RWMutex
-	workers       map[string]int
-	workersWg     sync.WaitGroup
-	stopRequested bool
+	runExecutor  RunExecutor
+	workers      map[models.ID]*singleJobSpecWorker
+	numWorkers   int32
+	chRuns       chan *models.JobRun
+	chStop       chan struct{}
+	chDone       chan struct{}
+	chWorkerDone chan models.ID
+}
 
-	runExecutor RunExecutor
+type singleJobSpecWorker struct {
+	chAddRun chan struct{}
+	chStop   chan struct{}
+	chDone   chan struct{}
 }
 
 // NewRunQueue initializes a RunQueue.
 func NewRunQueue(runExecutor RunExecutor) RunQueue {
 	return &runQueue{
-		workers:     make(map[string]int),
-		runExecutor: runExecutor,
+		workers:      make(map[models.ID]*singleJobSpecWorker),
+		runExecutor:  runExecutor,
+		chRuns:       make(chan *models.JobRun),
+		chStop:       make(chan struct{}),
+		chDone:       make(chan struct{}),
+		chWorkerDone: make(chan models.ID),
 	}
 }
 
 // Start prepares the job runner for accepting runs to execute.
 func (rq *runQueue) Start() error {
+	go rq.orchestrateWorkers()
 	return nil
 }
 
 // Stop closes all open worker channels.
 func (rq *runQueue) Stop() {
-	rq.workersMutex.Lock()
-	rq.stopRequested = true
-	rq.workersMutex.Unlock()
-	rq.workersWg.Wait()
+	close(rq.chStop)
+	<-rq.chDone
 }
 
 // Run tells the job runner to start executing a job
 func (rq *runQueue) Run(run *models.JobRun) {
-	rq.workersMutex.Lock()
-	if rq.stopRequested {
-		rq.workersMutex.Unlock()
-		return
+	select {
+	case rq.chRuns <- run:
+	case <-rq.chStop:
 	}
+}
+
+func (rq *runQueue) orchestrateWorkers() {
+	defer close(rq.chDone)
+	for {
+		select {
+		case run := <-rq.chRuns:
+			worker, exists := rq.workers[*run.ID]
+			if !exists {
+				worker = &singleJobSpecWorker{make(chan struct{}), make(chan struct{}), make(chan struct{})}
+				rq.workers[*run.ID] = worker
+				atomic.AddInt32(&rq.numWorkers, 1)
+				go rq.runSingleJobSpecWorker(run.ID, worker)
+				worker.chAddRun <- struct{}{}
+			} else {
+				select {
```

Comment:
Blocking the outer `select` is actually intentional here, as it avoids a race uncovered in the tests where the outer `select` can receive a message on `chStop` even though new work has already arrived (and therefore should be distributed to the worker).

Your concern about this sub-`select` blocking _should_ be entirely avoided by the fact that sending run requests on `worker.chAddRun` basically never blocks for more than a few nanoseconds.  Notice how the coordinator loop's `chAddRun` case simply increments `numRunsRequested` in the event that it can't send a message over `chResume`.  The worker can then pick up the queued runs as soon as it finishes its current workload.

Reply:
Ah, OK.